### ISSUE TITLE
Open Discussion Thread on plugin design

### ISSUE THREAD

**Issue by tonybaloney** [tonybaloney/tox-nuitka#1]
Thread to discuss how this plugin will work

## 1: User should be able to execute their tests using Nuitka

The goal of this plugin is to enable users of Tox to **also** run their test suite against the Nuitka compiler.

## 2: The default behaviour should execute commands with Nuitka

The default configuration should not require any runtime settings or flags, the user should fall into the "pit of success". 
TODO : __Determine defaults with Kay___

## 3: The user can test Nuitka on multiple versions of Python

The user may wish to test the Nuitka compiler on Python 2.7 and 3.x

## 4: The user can configure Nuitka CLI flags via the tox.ini configuration

Add config options via the plugin to set runtime and compilation flags for Nuitka

## 5: The user should still be able to execute Tox against CPython

* This is not an alternative test runner, this is an alternative execution mode for the tests.
* The user should still be able to test against CPython 2.7, 3.5, etc and PyPy

Cc @kayhayen

**Comment by kayhayen** [comment]
Yes, Nuitka should be on the level of uses implementation, although implicitely, behind Nuitka, there is always a Python version. So you can compile your code to become an extension module for Python 2.7 or 3.6, wouldn't matter.

**Comment by kayhayen** [comment]
I do not know much about tox, and what it does. But if it integrates with distutils, then the Nuitka one can be used. Ideally that needs no extra flag. I think on the command line, or with a boolean in the setup.py `setup ` call you can do this the compilation.

**Comment by kayhayen** [comment]
Test code if separate need not be compiled, actually that might even hide some issues with people later using resulting extension modules, but the code under test ought to be. If they are coupled, currently you have to compile both.

**Comment by kayhayen** [comment]
Ideally users need no flags. This is most probably not about standalone mode, where you want to disable recursion, but instead we want to compile the packages and modules provided in setup.

**Comment by kayhayen** [comment]
The documentation for the distutils stuff is not fully fleshed out yet. Consider this: https://github.com/kayhayen/Nuitka/blob/develop/tests/distutils/run_all.py

What it does is create a virtualenv and install with and without compilation. The "bdist_nuitka" is the target used, vs. normal "install".

**Comment by tonybaloney** [comment]
@kayhayen Tox essentially creates a virtualenv with a target version of Python, installs package dependencies within that venv and then runs the test commands for your project. It does this against multiple target runtime versions. You can also configure pre and post commands

**Comment by kayhayen** [comment]
My first question is, do you need anything beyond that, and does tox ultimately call things like .py.test, in which case it will need to something like this:

https://github.com/kayhayen/Nuitka/blob/develop/tests/test-runners/run_all.py

This compiles something, and makes sure there is a uncompiled file too and then calls py.test with it.

This is the raw material to work with. I didn't get "nose" to work, but I think "nose2" one time, but not so sure anymore.

**Comment by kayhayen** [comment]
@tonybaloney So basically it will run py.test in most instances directly, where does the software being tested come from, I assume it's not from PyPI, but pip installed locally or what?

**Comment by kayhayen** [comment]
@tonybaloney Checking the code, I see that it's not using setup.... maybe I am confused, but doesn't tox normally go through installing, is it normally bypassing that?

**Comment by obestwalter** [comment]
Hi @kayhayen, tox is very configurable and can run an arbitrary number of commands to prepare and run tests. 

If a project is not a package, setting `skipsdist = True` in the global section (`[tox]`)prevents building a package and `skip_install = True` in a `[testenv]` section will prevent installation of the package under test.

The [commands](https://tox.readthedocs.io/en/latest/config.html#confval-commands=ARGVLIST) section in a testenv can then run anything that is needed to run the tests.

Does that answer your questions?
